### PR TITLE
CSS: Reserve the use of the red color for links

### PR DESCRIPTION
**Pull request by tristanlatr**
Changes styles:

- functionHeader: bold and black
- fieldArg: black. Add a arrow before args
- margins: a little less, for more readability
- Also add a 35px margins to showPrivate toggle to able to click on "API Documentation" link at the bottom
- Make function's (source) links not bold
- merge the Sphinx build to be able to review the docs

Why?
Because the red color is usually for links and it needed some other CSS to mark function signature specials

### CODE REVIEW

**Comment by adiroiban** [comment]
Thanks for the update. It looks good. but first we need to merge #254 

I am happy that you used the code from #254 and I can review it via https://pydoctor--266.org.readthedocs.build/en/266/api/pydoctor.epydoc2stan.html 

So, if you have time, please send your feedback / review for PR #254 so that we can have it accepted.

Thanks again!

**Comment by tristanlatr** [comment]
The best would be to have the functions `def` lines generated with some kind of syntax highlight, like the class signature...

**Comment by mthuurne** [comment]
The `class` and `def` code fragments look much better in black.

I'm not a fan of the italics in the functions and variables definitions. I think they stand out enough already, making them italic as well just makes them harder to read. Also it means that items that were already italic such as "Undocumented" and "overrides" now stand out less.

For formatting the parameters, I think it's better to do a bigger overhaul in a separate PR in which we address #121.

To come back to the core change: instead of patching specific cases, maybe we should reconsider how we want to format code fragments in general. We have the following types of code fragments: (let me know if I missed any)
* definitions: `class X(Y):`, `def func():`, `var =`
* parameter names
* hyperlinked package/module, function, variable and type names
* inline docstring content marked up as code, for example `C{foo}` in epytext
* docstring content blocks, such as doctests

The current styling for code fragments consists of:
* font: monospaced at 90% size
* text color: red
* background: pink rounded rectangle

I think the font should be applied to all code fragments, since monospace is a standard way of hinting that something is code. Additionally, monospace fonts are often less ambiguous when it comes to for example "l vs I" (`l vs I`).

For the red text color, it may be best to reserve that for hyperlinked code. Having to hover over a fragment to know whether it is clickable is not a good user experience.

Regarding the background, I don't like the rectangles in the summary tables: when the background is already colored, having another colored rectangle on top of it makes it look cluttered. Maybe this could be fixed by tweaking the color of the code rectangle, but I think it's fine to just omit it, since most code fragments in summary tables are hyperlinked and therefore stand out already.

If we only make hyperlinked code fragments red, then no-link inline code fragments don't stand out very well. So having some kind of background on them is useful. For example, GitHub puts `code` on a light gray background. I assume the pink color was chosen to match the red text, so it would make sense to use gray when the text is black.

For inline code fragments that are hyperlinked, we could either keep the pink background, have a gray background or have no background. I've been playing around with "Inspect Element" and I think I prefer no background, but I'd have to see it in the full context of all the other changes to be sure.

> The best would be to have the functions `def` lines generated with some kind of syntax highlight, like the class signature...

Yes, it would be best to use the same CSS classes for definition lines as we use for syntax highlighting, since they are mimicking Python code.

**Comment by tristanlatr** [comment]
> I'm not a fan of the italics in the functions and variables definitions. I think they stand out enough already, making them italic as well just makes them harder to read. Also it means that items that were already italic such as "Undocumented" and "overrides" now stand out less.

No extra italics, I agree

> For example, GitHub puts code on a light gray background. I assume the pink color was chosen to match the red text, so it would make sense to use gray when the text is black.

Let's do as github no?

> For the red text color, it may be best to reserve that for hyperlinked code. Having to hover over a fragment to know whether it is clickable is not a good user experience.

I think it's addressed in the latests commits

This is now the 
The current styling for code fragments consists of:
font: monospaced at 90% size
text color: black
background: light gray

The current styling for links with code fragments consists of:
font: monospaced at 90% size
text color: red
background: pink rounded rectangle

+ I've added a border around a link to index.html that was transparent when using project url. It's not pretty but it works, It needs some other work, to address #121 and #257 anyway

:)

File: pydoctor/templates/apidocs.css [hunk 1]
```diff
@@ -18,7 +18,7 @@
 .categoryHeader {
     font-size: 24px;
     color: #777;
-    margin-bottom: 1.8em;
+    margin-bottom: 0.5em;
```
Comment (mthuurne) [inline pydoctor/templates/apidocs.css:21]: I like having a bit more margin between different parts of the document. Rather than reducing the margin here, we could increase margins elsewhere.

This is subjective of course, but in any case this change is not necessary to improve the code fragments layout.

File: pydoctor/templates/apidocs.css [hunk 2]
```diff
@@ -108,6 +108,7 @@ ul ul ul ul ul ul ul {
     padding: 5px;
 }
 
+
```
Comment (mthuurne) [inline pydoctor/templates/apidocs.css:111]: This extra empty line isn't consistent with the surrounding rules (a leftover of reverted edits?).

File: pydoctor/templates/apidocs.css [hunk 3]
```diff
@@ -157,8 +158,14 @@ tr.private {
 }
 
 .fieldArg {
-    font-weight: bold;
-    color: #c7254e;
+    color: #222222;
```
Comment (mthuurne) [inline pydoctor/templates/apidocs.css:161]: Do we need to override the text color here? The standard text color set by bootstrap is `#333333`, which doesn't look all that different from `#222222`.

**Comment by adiroiban** [comment]
There is a strange red border ... looks like an error

https://pydoctor--266.org.readthedocs.build/en/266/api/pydoctor.epydoc2stan.html

image

File: pydoctor/templates/apidocs.css [hunk 4]
```diff
@@ -157,8 +158,14 @@ tr.private {
...
+}
+
+/* Add an arrow before fields */
```
Comment (mthuurne) [inline pydoctor/templates/apidocs.css:164]: I prefer to handle the field layout in a separate PR, since we'll be looking at moving the type information to a different place as well.

File: pydoctor/templates/apidocs.css [hunk 5]
```diff
@@ -195,30 +202,90 @@ tr.private {
     margin-bottom: 6px;
 }
 
+/* This is magic to show the transparent index link */
```
Comment (mthuurne) [inline pydoctor/templates/apidocs.css:205]: I think it's useful to have a link to the index somewhere, but this is only half a solution, so I'd prefer to solve it completely in a separate PR.

File: pydoctor/templates/apidocs.css [hunk 6]
```diff
@@ -195,30 +202,90 @@ tr.private {
...
+
+/* .page-header > h1 > code > a {
+    background-color:#FCF9FA;
+} */
+
+/* 
+<a><code></code><a> for links
+This is inline docstring content marked up as code, for example C{foo} in epytext
+*/
+a > code {
```
Comment (mthuurne) [inline pydoctor/templates/apidocs.css:230]:
```suggestion
a[href] > code {
```
The `<a>` tag is also used to define anchors and that should not cause the text to become red.

File: pydoctor/templates/apidocs.css [hunk 7]
```diff
@@ -195,30 +202,90 @@ tr.private {
...
+    background-color:#f9f2f4;
+    border-radius: 4px;
+}
+
+/* This is for the main top navagation bar */
+.page-header > h1 > code {
+    background-color:#f9f2f4;
+    /* background-color:transparent; */
+}
+
```
Comment (mthuurne) [inline pydoctor/templates/apidocs.css:221]: There is some commented out code here (above and below) that can be cleaned up.

File: pydoctor/templates/apidocs.css [hunk 8]
```diff
@@ -195,30 +202,90 @@ tr.private {
...
+    margin-right: 0.3em;
+    border-width: 1.5px;
+    border-color: #c7254e;
+    background-color:#f9f2f4;
+    border-radius: 4px;
+}
+
+/* This is for the main top navagation bar */
+.page-header > h1 > code {
+    background-color:#f9f2f4;
```
Comment (mthuurne) [inline pydoctor/templates/apidocs.css:218]: Maybe we should attach the color to the `<h1>` tag rather than the `<code>` tag. Then we could later use the `package`/`module`/`class` styles that are currently applied to `<tr>` also on headers.

File: pydoctor/templates/apidocs.css (hunk 6 above)
Comment (mthuurne) [reply pydoctor/templates/apidocs.css:230]: Where do we use this rule? I'd expect most HTML would have the `<a href="">` inside a `<code>` tag and not vice versa.

File: pydoctor/templates/apidocs.css [hunk 9]
```diff
@@ -195,30 +202,90 @@ tr.private {
...
+*/
+a > code {
+    color:#c7254e; 
+    background-color:#f9f2f4;
+    /* background-color:transparent; */
+}
+/* 
+<a class="code"> links that are present in summary tables 
+*/
+a.code {
```
Comment (mthuurne) [inline pydoctor/templates/apidocs.css:238]: If we use the same styling in multiple places, we can use the `a, b { ... }` selector syntax to avoid duplicating the rules.

File: pydoctor/templates/apidocs.css [hunk 10]
```diff
@@ -195,30 +202,90 @@ tr.private {
...
+    /* color:#c7254e; 
+    background-color:#FCF9FA;*/
     border-radius:4px
 }
+/* 
+This overwrites the default values inherited from bootstrap min
+*/
+code {
+    color: #222222;
+    background-color: #f9f2f4;
```
Comment (mthuurne) [inline pydoctor/templates/apidocs.css:261]: I think it would look better if we used pink background for red text and gray background for near-black text.

File: pydoctor/templates/apidocs.css [hunk 11]
```diff
@@ -195,30 +202,90 @@ tr.private {
...
 
 code > .code {
     padding: 0px;
 }
 
+
+
+a.functionSourceLink {
+    font-weight: normal;
+}
```
Comment (mthuurne) [inline pydoctor/templates/apidocs.css:272]: This is good for now. Later we should remove the bold from `functionHeader` and instead style the function definition line like syntax highlighted code.

File: pydoctor/templates/apidocs.css [hunk 12]
```diff
@@ -195,30 +202,90 @@ tr.private {
...
+
+
 #childList > div {
     border-left-color: #03a9f4;
     border-left-width: 1px;
     border-left-style: solid;
-    background: #fafafa;
+    background: #fbfbfb;
```
Comment (mthuurne) [inline pydoctor/templates/apidocs.css:279]: That's a very subtle change ;)

File: pydoctor/templates/apidocs.css [hunk 13]
```diff
@@ -250,6 +317,7 @@ body.private-hidden #splitTables .private, body.private-hidden #childList .priva
     left: 0;
     padding: 10px;
     opacity: 0.7;
+    margin-bottom: 35px;
```
Comment (mthuurne) [inline pydoctor/templates/apidocs.css:320]: This seems unrelated to the other changes.

**Review by mthuurne — CHANGES_REQUESTED** [review]
To summarize:
- overall this looks much better :)
- I would like changes that are unrelated to styling code fragments to be moved to separate PRs
- the CSS still has several leftovers from experiments that need to be cleaned up before we can merge

**Comment by mthuurne** [comment]
By the way, did you try disabling the background rectangle for code in the summary tables? If I do that using the browser's dev mode, I think the result is easier on the eyes.

File: pydoctor/templates/apidocs.css (hunk 8 above)
Comment (tristanlatr) [reply pydoctor/templates/apidocs.css:218]: With `.page-header > h1 > * {` instead of `.page-header > h1 > code {` it works

File: pydoctor/templates/apidocs.css (hunk 6 above)
Comment (tristanlatr) [reply pydoctor/templates/apidocs.css:230]: > Where do we use this rule?

Here for instance: file:///Users/tristan/Documents/pydoctor/build/apidocs/pydoctor.epydoc.markup.ParsedDocstring.html

The `Subclasses must implement L{to_stan()}.`

File: pydoctor/templates/apidocs.css (hunk 8 above)
Comment (mthuurne) [reply pydoctor/templates/apidocs.css:218]: Hmm, not sure that's an improvement. I was thinking of using just `.page-header > h1 {`, but it seems quite a lot of CSS needs to change for that to work.

Also the `<h1>` spans the entire body width instead of just the title, so it would look differently.

Let's leave it as- is, for now.

File: pydoctor/templates/apidocs.css (hunk 6 above)
Comment (mthuurne) [reply pydoctor/templates/apidocs.css:230]: Hmm, we could change the code that outputs this in `pydoctor.epydoc.markup.*`, but I'm not sure whether that would make it easier or harder to write the CSS.

**Comment by tristanlatr** [comment]
> There is a strange red border ... looks like an error

That was't an error, but a link to the index

> By the way, did you try disabling the background rectangle for code in the summary tables? If I do that using the browser's dev mode, I think the result is easier on the eyes.

Yes, it was the `/* background-color:transparent; */`. I find it cool too. 

The requested changes are done.

**Comment by mthuurne** [comment]
Thanks!

The command like `git` client shows some redundant spaces or newline characters at the end of some changed lines, so I added a commit to remove those.
